sqlmesh/core/context.py

```diff
@@ -1597,53 +1599,156 @@ def table_diff(
-            ).qualified_view_name.for_environment(target_env.naming_info, adapter.dialect)
-
-            source_alias = source_env.name
-            target_alias = target_env.name
-
-            if not on:
-                on = []
-                for expr in [ref.expression for ref in model.all_references if ref.unique]:
-                    if isinstance(expr, exp.Tuple):
-                        on.extend(
-                            [key.this.sql(dialect=adapter.dialect) for key in expr.expressions]
-                        )
-                    else:
-                        # Handle a single Column or Paren expression
-                        on.append(expr.this.sql(dialect=adapter.dialect))
+                raise SQLMeshError(f"Could not find environment '{target}'")
+
+            modified_snapshots: t.Set[ModelOrSnapshot] = (
+                {model_or_snapshot} if model_or_snapshot else set()
+            )
+            if select_model:
+                models_to_diff = self._new_selector().expand_model_selections(select_model)
+                target_snapshots = {
+                    s.name: s
+                    for s in self.state_reader.get_snapshots(target_env.snapshots).values()
+                    if s.name in models_to_diff
+                }
+                context_diff = self._context_diff(
+                    source,
+                    snapshots=target_snapshots,
+                    ensure_finalized_snapshots=self.config.plan.use_finalized_state,
+                )
+                modified_snapshots = {
+                    current_snapshot.snapshot_id.name
+                    for _, (current_snapshot, _) in context_diff.modified_snapshots.items()
+                }
+            tasks_num = min(len(modified_snapshots), self.concurrent_tasks)
+            table_diffs = concurrent_apply_to_values(
+                list(modified_snapshots),
+                lambda s: self._model_diff(
+                    source_env=source_env,
+                    target_env=target_env,
+                    model_or_snapshot=s,
+                    limit=limit,
+                    decimals=decimals,
+                    on=on,
+                    skip_columns=skip_columns,
+                    where=where,
+                    show=show,
+                    temp_schema=temp_schema,
+                    skip_grain_check=skip_grain_check,
+                ),
+                tasks_num=tasks_num,
+            )
+        else:
+            table_diffs = [
+                self._table_diff(
+                    source=source,
+                    target=target,
+                    source_alias=source,
+                    target_alias=target,
+                    limit=limit,
+                    decimals=decimals,
+                    adapter=self.engine_adapter,
+                    on=on,
+                    skip_columns=skip_columns,
+                    where=where,
+                )
+            ]
+
+        if show:
```

Comment:
I think this should be handled in the CLI code and the `show` argument removed from this method.